### ISSUE TITLE
Traincascade: grabbing negative window samples is seriously broken

### ISSUE THREAD

**Issue by StevenPuttemans** [opencv/opencv#6520]
### Please state the information for your system
- OpenCV version: 3.x --> latest master branch
- Host OS: Linux (Ubuntu 14.04) 

### In which part of the OpenCV library you got the issue?
apps --> opencv_traincascade process

### Expected behaviour
Providing more then enough negative samples is a common approach in computer vision. This makes sure that through the rescaling and subsampling windows, the train cascade application succeeds in grabbing multiple millions of windows from a set of larger images.

However this is not the case IF you supply a limited set of data.

### Actual behaviour & problem description

Imagine the following setup (just confirmed on a detector setup I have here)

- Given 2500 positive windows of 19x14 pixels
- Given 1 negative image of 66x28 pixels, from which negative windows can be grabbed
- Training LBP feature cascade classifier with `-numStages 25 -w 19 -h 14`

### Generated output

```
spu@TOBCAT:/data/zonnepanelen_test$ opencv_traincascade -data cascade/ -vec data.vec -bg negatives.txt -numPos 2500 -numNeg 10000 -numStages 25 -precalcValBufSize 8000 -precalcIdxBufSize 8000 -w 19 -h 14 -featureType LBP -minHitRate 0.999
PARAMETERS:
cascadeDirName: cascade/
vecFileName: data.vec
bgFileName: negatives.txt
numPos: 2500
numNeg: 10000
numStages: 25
precalcValBufSize[Mb] : 8000
precalcIdxBufSize[Mb] : 8000
acceptanceRatioBreakValue : -1
stageType: BOOST
featureType: LBP
sampleWidth: 19
sampleHeight: 14
boostType: GAB
minHitRate: 0.999
maxFalseAlarmRate: 0.5
weightTrimRate: 0.95
maxDepth: 1
maxWeakCount: 100
Number of unique features given windowSize [19,14] : 1710

===== TRAINING 0-stage =====
<BEGIN
POS count : consumed   2500 : 2500
NEG count : acceptanceRatio    10000 : 1
Precalculation time: 0
+----+---------+---------+
|  N |    HR   |    FA   |
+----+---------+---------+
|   1|        1|        1|
+----+---------+---------+
|   2|        1|        1|
+----+---------+---------+
|   3|        1|        1|
+----+---------+---------+
|   4|   0.9996|   0.1268|
+----+---------+---------+
END>
Training until now has taken 0 days 0 hours 0 minutes 3 seconds.

===== TRAINING 1-stage =====
<BEGIN
POS count : consumed   2500 : 2501
NEG count : acceptanceRatio    10000 : 0.127781
Precalculation time: 0
+----+---------+---------+
|  N |    HR   |    FA   |
+----+---------+---------+
|   1|        1|        1|
+----+---------+---------+
|   2|        1|        1|
+----+---------+---------+
|   3|        1|        1|
+----+---------+---------+
|   4|   0.9996|   0.0994|
+----+---------+---------+
END>
Training until now has taken 0 days 0 hours 0 minutes 8 seconds.

===== TRAINING 2-stage =====
<BEGIN
POS count : consumed   2500 : 2502
NEG count : acceptanceRatio    10000 : 0.0126893
Precalculation time: 0
+----+---------+---------+
|  N |    HR   |    FA   |
+----+---------+---------+
|   1|        1|        1|
+----+---------+---------+
|   2|   0.9992|        0|
+----+---------+---------+
END>
Training until now has taken 0 days 0 hours 0 minutes 33 seconds.

===== TRAINING 3-stage =====
<BEGIN
POS count : consumed   2500 : 2504
```

### Remarks on the generated output

1. How is it possible that from a single window of 66x18 pixels, more then 3 stages of 10.000 negatives can be grabbed? Imagining that the second stages only takes negatives that are not classified as negatives by the previous cascade, this can grow easily up to 40.000 - 50.000 samples just for those 3 stages trained. 
2. A known problem is that the training hangs once it cannot find new negative samples. But in my opinion in this case it hangs way to late, and should rather drop an error message saying, `oh hello, I cannot find any more negative windows, do provide me more if you want to continue training`

**Comment by sjelavic** [comment]
Hi @mshabunin I also had a similar issue. I create a single stage classifier using 300 identical negative images and low FA. If I run the detection of that same negative image using the generated classifier I get 26 false positives. Something definitely wrong with this code.

**Comment by StevenPuttemans** [comment]
@sjelavic I agree that you should not get the suggested output, but, why do you supply 300 identical negative images? There is no difference in supplying a single one. Can you provide your complete training set parameters?

It is perfectly possible that with only a single unique sample, to generate a model that still has 26 FP detections ... so I am pretty sure this has nothing to do with this current issue in the code.

**Comment by sjelavic** [comment]
@StevenPuttemans Use 300 identical samples because there are constraints on how many min neg samples you can have. Not sure if I understand your 26 FP statement. How is it possible to have 26 falses if training with the same image gives a low false rate. It is the same process used during training and detection.

**Comment by StevenPuttemans** [comment]
@sjelavic actually that is were you are wrong
- the reported FA rate is during training on the given training samples
- the FPs during detection are on test data for which the model id not optimized

So you need to differentiate your training data to cope with class specific variance

**Comment by LorenaGdL** [comment]
@sjelavic I guess you're misunderstanding something: even though you may provide 300 identical negative images to the app, the training procedure samples negative windows from each of the images, so your negative subset is composed by images which are not identical (though it will be repeated negative windows, as the extraction is sequential and fixed).

**Comment by StevenPuttemans** [comment]
@LorenaGdL I was guessing that with negative images he mentioned like hard negatives so window sizes, but indeed if it is larger there will be sampling.

**Comment by LorenaGdL** [comment]
> If I run the detection of that same negative image using the generated classifier I get 26 false positives

You can't have more than one prediction if your image is of window size, can you? ;)

**Comment by sjelavic** [comment]
@StevenPuttemans and @LorenaGdL  I thought sampling of negative windows is a part of the detection process that is used to determine if there is a false detect which is the same procedure used during actual object detection. Are you saying that the detection procedure is not the same during training and during detect? To think of it, why would the detect process be different between training and detection? In another words, the training process should have detection that is equivalent to the actual object detection, otherwise the FA and HR do not compare.

**Comment by StevenPuttemans** [comment]
Wow wait, there is a very difference between training a model and using it to detect objects. I am getting the feeling that you do not understand that.

Sampling of negative windows happens from negative images provided in the `-bg` file provided to the `opencv_traincascade` application. The interface takes windows the size of the actual object model, and goes in a sliding window manner over the negative provided background images to provide negative windows, which is specified by the `-numNeg` parameter.

During detection there is no longer the collection of data, but rather each window is evaluated using the features of the model and using the cascade waterfall principle in combination with early rejection you decide whether a sample is positive or negative. 

More information in the basic Viola and Jones paper: https://www.cs.cmu.edu/~efros/courses/LBMV07/Papers/viola-cvpr-01.pdf

**Comment by StevenPuttemans** [comment]
And NO, FA and HR do not compare in both situations

- During training you specify the minimal HR on the positive training samples and then form a weak classifier stage based on features until your FA of that stage drops below the set FA rate, 0.5 by default.

- During detection you evaluate on unseen test data, so yes the HR and FA ratio defined during training make 0 sense at all

**Comment by sjelavic** [comment]
I see that you miss-understand me @StevenPuttemans .  You say:
During detection you evaluate on unseen test data, so yes the HR and FA ratio defined during training make 0 sense at all

Not true. because I use the SAME image during training and detection.

**Comment by StevenPuttemans** [comment]
Can you then, I ask again, provide the data used? Because I have a hard time believing that with a single input image, given the same input image you are not able to reduce the FP more... training data should closely model your trained model...

**Comment by sjelavic** [comment]
I will send you the data. I will have to redo it because some of it I cannot share.

**Comment by sjelavic** [comment]
Here is the data @StevenPuttemans 
Enclosed are positive and negative images and the detect output showing falses.
I ran the following training commands.

perl ../bin/createsamples.pl positives.txt negatives.txt samples 10 "createsamples.exe -bgcolor 210 -bgthresh 45 -maxxangle 1.1 -maxyangle 1.1 maxzangle 0.5 -maxidev 40 -show 4.0 -w 20 -h 20"

./traincascade.exe -data classifier -vec samples.vec -bg negatives.txt -numStages 10 -minHitRate 0.999 -maxFalseAlarmRate 0.001 -numPos 3 -numNeg 10 -w 20 -h 20 -mode ALL -precalcValBufSize 1024 -precalcIdxBufSize 1024

![pen_falsing](https://cloud.githubusercontent.com/assets/16308308/15578531/cee036e0-232e-11e6-823c-4631eff5e230.jpg)
![1](https://cloud.githubusercontent.com/assets/16308308/15578555/e1d0c22e-232e-11e6-8155-ee75ad72aede.jpg)
![1](https://cloud.githubusercontent.com/assets/16308308/15578562/e87c0732-232e-11e6-9487-97e1d5043ffb.jpg)

**Comment by StevenPuttemans** [comment]
Ow man, please go read all the available info on cascade classifier training on the website http://answers.opencv.org/questions especially the parameter explanation parts...

Your mistake is obvious, you are resizing training samples to -w -h 20 which is a square object. On the contrary a pen does not have those dimensions.

Also a FA for a single stage at 0.001 makes no sense at and wont build you weak classifier...

**Comment by sjelavic** [comment]
Ohhh man. OK here is the corrected set of data but I still have that same issue. Positive image (pen) size is 20x20 now. The negative is the image with the telephone 300x300 pixels. Here is the training output:

PARAMETERS:
cascadeDirName: classifier
vecFileName: samples.vec
bgFileName: negatives.txt
numPos: 3
numNeg: 10
numStages: 1
precalcValBufSize[Mb] : 1024
precalcIdxBufSize[Mb] : 1024
stageType: BOOST
featureType: HAAR
sampleWidth: 20
sampleHeight: 20
boostType: GAB
minHitRate: 0.999
maxFalseAlarmRate: 0.5
weightTrimRate: 0.95
maxDepth: 1
maxWeakCount: 100
mode: ALL

===== TRAINING 0-stage =====
<BEGIN
POS count : consumed   3 : 3
NEG count : acceptanceRatio    10 : 1
Precalculation time: 0.016
+----+---------+---------+
|  N |    HR   |    FA   |
+----+---------+---------+
|   1|        1|        0|
+----+---------+---------+
END>
Training until now has taken 0 days 0 hours 0 minutes 0 seconds.

Now if I take that negative image and run object detect software using the classifier generated  and  I see false detects all over the place. 

![falsing1](https://cloud.githubusercontent.com/assets/16308308/15589886/c67f1258-2362-11e6-9671-41f3e61e99da.jpg)

**Comment by StevenPuttemans** [comment]
Please reread my comment... a pen does not have a 1:1 ratio in real life so why the hell would you train a 20x20 sized model. In addition to that, your model consists of a single stage with a single weak classifier stump evaluating a single feature able to seperate the given negative from the positive. Your negative gets rescaled which leads mainly to this behaviour...

But again there is tons of stuff plain wrong here ...

**Comment by LorenaGdL** [comment]
@sjelavic onece again, negative samples are grabbed from the provided negative image. You set to use 10 negative windows per stage, but your original 300x300 negative image contains several 20x20 negative windows. Only the first ten are used and successfully discarded (in this case, as FA = 0), but still your image has plenty of 20x20 windows that may be classified as false positives.

**Comment by StevenPuttemans** [comment]
@LorenaGdL thank you for explaining it more clearly

**Comment by sjelavic** [comment]
Thank you @LaurentBerger . That is what I was looking for. Now my 1+1 adds to 2 :)

**Comment by StevenPuttemans** [comment]
@alalek is there a way of cleaning this up? The conversation done here is actually unrelated to the initial issue posted and the initial issue is still available in code and should be fixed. Therefore the removed labels should be placed back or a new issue should be opened to keep track of this and this one closed down.

**Comment by StevenPuttemans** [comment]
As I am the author, I can provide it myself.
